### PR TITLE
IGNITE-22745 DB API Driver 3: Provide wider data types support

### PR DESCRIPTION
**Pull request by isapego**
https://issues.apache.org/jira/browse/IGNITE-22745

- Implemented values reading;
- Implemented values writing;
- Re-factored and added some new tests;
- Added conversion functions.

### CODE REVIEW

File: modules/platforms/python/cpp_module/utils.cpp [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+    auto kwargs = PyDict_New();
+    if (!kwargs)
+        return nullptr;
+    auto kwargs_guard = ignite::detail::defer([&]{ Py_DECREF(kwargs); });
+
+    return PyObject_Call(datetime_class, args, kwargs);
+}
+
+PyObject* py_create_datetime(const ignite::ignite_timestamp &value) {
+    // TODO: Cache classes and functions for re-use
```
Comment (ptupitsyn) [inline modules/platforms/python/cpp_module/utils.cpp:313]: Ticket for TODO?

File: modules/platforms/python/cpp_module/type_conversion.h [hunk 2]
```diff
@@ -171,8 +332,39 @@ static void submit_pyobject(ignite::binary_tuple_builder &builder, PyObject *obj
...
+            ignite::protocol::append_type_and_scale(builder, ignite::ignite_type::BOOLEAN);
+            builder.append_bool(val);
+        }
+        return;
+    }
+
     if (PyFloat_Check(obj)) {
         double val = PyFloat_AsDouble(obj);
+        if (PyErr_Occurred()) {
+            throw ignite::ignite_error("Can not convert FLOAT to a double");
```
Comment (ptupitsyn) [inline modules/platforms/python/cpp_module/type_conversion.h:365]: How is that possible?

File: modules/platforms/python/cpp_module/type_conversion.h [hunk 3]
```diff
@@ -185,6 +377,10 @@ static void submit_pyobject(ignite::binary_tuple_builder &builder, PyObject *obj
 
     if (PyLong_Check(obj)) {
         auto val = PyLong_AsLongLong(obj);
+        if (PyErr_Occurred()) {
+            throw ignite::ignite_error("Can not convert INT to a long long");
```
Comment (ptupitsyn) [inline modules/platforms/python/cpp_module/type_conversion.h:381]: Same as above - is that possible in practice?

File: modules/platforms/python/cpp_module/type_conversion.h [hunk 4]
```diff
@@ -85,23 +88,59 @@ static PyObject* primitive_to_pyobject(ignite::primitive value) {
...
+            return py_create_number(str);
+        }
+
+        case ignite_type::DURATION: {
+            auto &duration_val = value.get<ignite::ignite_duration>();
+            return py_create_timedelta(duration_val);
+        }
+
+        case ignite_type::PERIOD:{
+            PyErr_SetString(PyExc_RuntimeError, "PERIOD data type is not supported");
```
Comment (ptupitsyn) [inline modules/platforms/python/cpp_module/type_conversion.h:138]: Do we need a ticket for this?

File: modules/platforms/python/cpp_module/type_conversion.h [hunk 5]
```diff
@@ -122,40 +161,162 @@ static void submit_pyobject(ignite::binary_tuple_builder &builder, PyObject *obj
         return;
     }
 
-    if (PyBool_Check(obj)) {
-        bool val = (obj == Py_True);
+    if (PyObject_IsInstance(obj, py_get_module_timestamp_class())) {
+        auto double_value = PyFloat_AsDouble(obj);
+        if (PyErr_Occurred()) {
+            throw ignite::ignite_error("Can not get ticks count from a TIMESTAMP value");
```
Comment (ptupitsyn) [inline modules/platforms/python/cpp_module/type_conversion.h:167]: Is that possible in practice? Should we add the original error message from `PyErr_Occurred`?

File: modules/platforms/python/cpp_module/type_conversion.h [hunk 6]
```diff
@@ -122,40 +161,162 @@ static void submit_pyobject(ignite::binary_tuple_builder &builder, PyObject *obj
...
     }
 
-    if (PyBytes_Check(obj)) {
-        auto *data = reinterpret_cast<std::byte*>(PyBytes_AsString(obj));
-        auto len = PyBytes_Size(obj);
-        ignite::bytes_view view(data, len);
+    if (PyObject_IsInstance(obj, py_get_module_number_class())) {
+        auto obj_str = PyObject_Str(obj);
+        if (!obj_str) {
+            throw ignite::ignite_error("Can not convert NUMBER to str");
```
Comment (ptupitsyn) [inline modules/platforms/python/cpp_module/type_conversion.h:186]: Here and other places - if a conversion fails, we should provide more details:
- The value that we could not convert
- The original error from converter, if applicable

**Review by ptupitsyn — CHANGES_REQUESTED** [review]

File: modules/platforms/python/cpp_module/type_conversion.h (hunk 2 above)
Comment (isapego) [reply modules/platforms/python/cpp_module/type_conversion.h:365]: In my understanding, it should not be possible, but this is not clarified in the documentation, so I just decided to play safe.

File: modules/platforms/python/cpp_module/utils.cpp (hunk 1 above)
Comment (isapego) [reply modules/platforms/python/cpp_module/utils.cpp:313]: Good catch. Fixed it already, just forgot to remove TODO.

File: modules/platforms/python/cpp_module/type_conversion.h (hunk 3 above)
Comment (isapego) [reply modules/platforms/python/cpp_module/type_conversion.h:381]: This is actually possible, when a really big int is used in Python (e.g., I tried `423538409739423538409739423538409739423538409739423538409739423538409739423538409739` and got this error)

File: modules/platforms/python/cpp_module/type_conversion.h (hunk 4 above)
Comment (isapego) [reply modules/platforms/python/cpp_module/type_conversion.h:138]: Probably. The problem is there is no type like this in Python. I'll file a ticket.

File: modules/platforms/python/cpp_module/type_conversion.h (hunk 2 above)
Comment (isapego) [reply modules/platforms/python/cpp_module/type_conversion.h:365]: Ok, I've checked the sources and found that if the object is Double, we [should not do any checks](https://github.com/python/cpython/blob/14e5bdceff45e6e789e1f838b96988946c75b0f4/Objects/floatobject.c#L262). Changed to `PyFloat_AS_DOUBLE`.

File: modules/platforms/python/cpp_module/type_conversion.h (hunk 4 above)
Comment (isapego) [reply modules/platforms/python/cpp_module/type_conversion.h:138]: https://issues.apache.org/jira/browse/IGNITE-23217

File: modules/platforms/python/cpp_module/type_conversion.h (hunk 5 above)
Comment (isapego) [reply modules/platforms/python/cpp_module/type_conversion.h:167]: It's not, I'll handle it just as a float type. Added getting an error message from an exception where possible.

**Review by ptupitsyn — APPROVED** [review]

**Pull request merged**